types/types.cc

```diff
@@ -844,6 +846,45 @@ bool abstract_type::references_user_type(const sstring& keyspace, const bytes& n
     return find(*this, visitor{keyspace, name});
 }
 
+namespace {
+// For tuples and collection types: recurse over inner types
+// For user-defined types: add it's name to referenced udts
+// For other types: do nothing
+struct get_all_referenced_user_types_visitor {
+    std::unordered_set<sstring> referenced_udts;
+
+    template<typename T>
+    void operator()(const concrete_type<T>&) { return; }
+    void operator()(const counter_type_impl&) { return; }
+    void operator()(const empty_type_impl&) { return; }
+    void operator()(const reversed_type_impl& r) {
+        visit(*r.underlying_type(), *this);
+    }
+    void operator()(const user_type_impl& u) {
+        referenced_udts.insert(u.get_name_as_string());
+    }
+    void operator()(const tuple_type_impl& t) {
+        for (auto& field: t.all_types()) {
+            visit(*field, *this);
+        }
+    }
+    void operator()(const map_type_impl& m) {
+        visit(*m.get_keys_type(), *this);
+        visit(*m.get_values_type(), *this);
+    }
+    void operator()(const listlike_collection_type_impl& l) {
+        visit(*l.get_elements_type(), *this);
+    }
```

Comment:
ok, so now we don't have a default and instead have a few null cases for leaf types. I think it's an improvement.